Better PHP8 type handling

adambalint-srg opened the pull request:
I've found two type declaration related problems in proxy generator, when using it with PHP8.

First is the usage of static keyword, which should be "copied" to the proxy class, and also can be nullable.
`function setId(int $id): ?static`
This is supported in this pull request.

Second is a fix for union types of PHP8, because in case of nullable union types the '?' sign is nut supported. It should be used in this way:
`float|int|null`
But the proxy generator prepended the '?' to the null part of the union type:
`float|int|?null`
This problem has been fixed with an optional argument in case of recursive call of formatType function.

adambalint-srg committed:
Fix nullable union types for PHP8

In case of nullable union types, '?' sign is not supported. The current solution for uniont types (the recursive call) prepended the '?' sign to the null type, so I had to handle this with an additional option for the formatType method of the ProxyGenerator.

adambalint-srg committed:
Support for PHP8 static return value in proxies

PHP8 supports 'static' keyword as a return type. This should be copied to the proxy class without any transformation. This type is not a "builtin" type in php, so I should extend some if branches with an additional condition so allow this new keyword.

adambalint-srg committed:
Cody style check fixes (tabs/spaces)

jellynoone commented on lib/Doctrine/Common/Proxy/ProxyGenerator.php:
```diff
@@ -1151,12 +1156,13 @@ function (ReflectionType $unionedType) use ($method, $parameter) {
 
-        if (! $type->isBuiltin()) {
+        if (! $type->isBuiltin() && $name !== 'static') {
             $name = '\\' . $name;
         }
 
         if (
-            $type->allowsNull()
+            ! $isUnionPart
+            && $type->allowsNull()
             && ($parameter === null || ! $parameter->isDefaultValueAvailable() || $parameter->getDefaultValue() !== null)
             && $name !== 'mixed'
```
Thanks for working on this just found the nullable issue myself. 

Would it perhaps be better to avoid creating a new parameter and just check for name null like it is done for mixed?

```php
if (
    $type->allowsNull()
    && ! in_array($name, ['mixed', 'null'], true)
    && ($parameter === null || ! $parameter->isDefaultValueAvailable() || $parameter->getDefaultValue() !== null)
) 
```

adambalint-srg replied:
Thanks for the suggestion, it's more logical :) I've fixed this with a new commit

adambalint-srg committed:
Modify nullable check without additional function parameter

With a name check it's more logical to check the null value, because this is the only one when the unnecessary '?' sign can occure. So no function parameter is required with this solution. Thanks @jellynoone

jellynoone commented:
Can someone from the Doctrine team review the pull request? @beberlei  @greg0ire 

By now the author, myself and https://github.com/doctrine/orm/issues/8422 have found the issue being solved in this pull request.

beberlei commented:
Does this target the 3.1.x branch for a specific reason? This is a bugfix for me and should target 3.0.x

Edit: Sorry, my mistake Github showed 3.0.3 as last release, but it was actually 3.1.0 already.
